### PR TITLE
Optimize Shape related computation subgraphs

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR adds a graph rewrite rule to avoid redundant compute preceding a Shape op. It also adds the TransposeOfShape op that returns the transposed shape of the input according to the specified perm.
For example, :
1)
Cast -> Shape
Here, if the Cast has no other consumers, it is Redundant.

2)
Transpose -> Shape
If the Transpose has no other consumers, we can transpose the shape instead of 
transposing the larger-sized activation.
TransposeOfShape

Similar patterns can be added to this Rewrite Rule as required.

### CODE REVIEW

File: onnxruntime/core/optimizer/shape_optimization.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+For example, :
+1)
+Cast -> Shape
+Here, if the Cast has no other consumers, it is Redundant.
+
+2)
+Transpose -> Shape
+If the Transpose has no other consumers, we can transpose the shape instead of 
+transposing the larger-sized activation.
+Shape -> Transpose
```
Comment (ashbhandare) [inline onnxruntime/core/optimizer/shape_optimization.h:23]: change to TransposeOfShape

File: onnxruntime/core/optimizer/shape_optimization.cc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      graph.RemoveEdge(cast_input_node->Index(), cast_node.Index(), output_idx, 0);
+    }
+    graph.RemoveEdge(cast_node.Index(), shape_node.Index(), 0, 0);
+    graph_utils::ReplaceNodeInput(shape_node, 0, *new_input_arg_for_shape);
+
+    if (cast_input_node) {
+      graph.AddEdge(cast_input_node->Index(), shape_node.Index(), output_idx, 0);
+    }
+    // remove cast_node
+    graph.RemoveNode(cast_node.Index());
```
Comment (ytaous) [inline onnxruntime/core/optimizer/shape_optimization.cc:34]: >RemoveNode [](http://example.com/codeflow?start=10&length=10)

would this be sufficient, thus we don't need the 2 remove edges above?

File: onnxruntime/core/optimizer/shape_optimization.cc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    // Assign provider to this new node. Provider should be same as the provider for old node.
+    fused_node.SetExecutionProviderType(shape_node.GetExecutionProviderType());
+
+    // move the output definition and edges from the mul_node to the div_node and delete the mul_node
+    graph_utils::FinalizeNodeFusion(graph, {transpose_node, shape_node}, fused_node);
+
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+
+#else
+    rule_effect = RewriteRuleEffect::kNone;
```
Comment (ytaous) [inline onnxruntime/core/optimizer/shape_optimization.cc:63]: >kNone [](http://example.com/codeflow?start=37&length=5)

can move this to top as default, and remove both kNone here

File: onnxruntime/core/optimizer/shape_optimization.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    graph.RemoveNode(cast_node.Index());
+
+    rule_effect = RewriteRuleEffect::kUpdatedCurrentNode;
+
+  } else if (op_type == "Transpose") {
+#ifndef DISABLE_CONTRIB_OPS
+    auto& transpose_node = previous_node;
+    // add TransposeOfShape node
+    const std::vector<NodeArg*> new_input_defs{transpose_node.MutableInputDefs()[0]};
+    Node& fused_node = graph.AddNode(graph.GenerateNodeName("TransposeOfShape"),
```
Comment (ytaous) [inline onnxruntime/core/optimizer/shape_optimization.cc:43]: >TransposeOfShape [](http://example.com/codeflow?start=61&length=16)

should we add the original transpose name here?

File: onnxruntime/core/optimizer/shape_optimization.cc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#include "core/graph/op.h"
+#include "core/optimizer/initializer.h"
+
+namespace onnxruntime {
+
+Status ShapeOptimization::Apply(Graph& graph, Node& shape_node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  Node& previous_node = *graph.GetNode(shape_node.InputNodesBegin()->Index());
+
+  auto op_type = previous_node.OpType();
+  if (op_type == "Cast") {
```
Comment (ytaous) [inline onnxruntime/core/optimizer/shape_optimization.cc:16]: >Cast [](http://example.com/codeflow?start=18&length=4)

should we move this cast elimination to CastElimination ?

File: onnxruntime/core/providers/cpu/tensor/shape_op.h [hunk 6]
```diff
@@ -67,4 +67,58 @@ class Shape final : public OpKernel {
...
+  TransposeOfShape(const OpKernelInfo& info) : OpKernel(info) {
+    std::vector<int64_t> temp_perm;
+    Status status = info.GetAttrs<int64_t>("perm", temp_perm);
+    if (status.IsOK()) {
+      size_t rank = temp_perm.size();
+      perm_.resize(temp_perm.size());
+      // Check that perm_ is a valid permutation of [0,rank-1]
+      for (size_t i = 0; i != temp_perm.size(); ++i) {
+        int64_t v = temp_perm[i];
+        ORT_ENFORCE(v >= 0 && static_cast<uint64_t>(v) <= std::numeric_limits<size_t>::max());
```
Comment (ytaous) [inline onnxruntime/core/providers/cpu/tensor/shape_op.h:81]: >v [](http://example.com/codeflow?start=52&length=1)

SafeInt? so you don't need this

File: onnxruntime/core/providers/cpu/tensor/shape_op.h [hunk 7]
```diff
@@ -67,4 +67,58 @@ class Shape final : public OpKernel {
...
+  // containing the shape of the input tensor.
+  Status Compute(OpKernelContext* context) const override {
+    const auto* input = context->Input<Tensor>(0);
+    const TensorShape& input_shape = input->Shape();
+
+    size_t rank = input_shape.NumDimensions();
+    Tensor* output = context->Output(0, {gsl::narrow_cast<int64_t>(rank)});
+    int64_t* output_data = output->template MutableData<int64_t>();
+
+    if (perm_specified_) {  // vanilla use of Shape (no slicing)
```
Comment (ytaous) [inline onnxruntime/core/providers/cpu/tensor/shape_op.h:106]: >slicing [](http://example.com/codeflow?start=56&length=7)

invalid comments?

**Pull request closed without merging**